rust/lance-index/src/scalar/ngram.rs

```diff
@@ -0,0 +1,456 @@
+            );
+        }
+        Ok(frag_ids)
+    }
+}
+
+#[async_trait]
+impl ScalarIndex for NGramIndex {
+    async fn search(&self, query: &dyn AnyQuery) -> Result<SearchResult> {
+        let query =
+            query
+                .as_any()
+                .downcast_ref::<TextQuery>()
+                .ok_or_else(|| Error::InvalidInput {
+                    source: "Query is not a TextQuery".into(),
+                    location: location!(),
+                })?;
+        match query {
+            TextQuery::StringContains(substr) => {
+                let tokens = self.tokenize(substr);
+                let token_ids = tokens
+                    .into_iter()
+                    .map(|t| self.tokens.get(&t))
+                    .collect::<Option<Vec<_>>>();
+                let Some(token_ids) = token_ids else {
+                    return Ok(SearchResult::Exact(RowIdTreeMap::new()));
+                };
+                let posting_lists = futures::stream::iter(
+                    token_ids
+                        .into_iter()
+                        .map(|token_id| self.list_reader.ngram_list(token_id)),
+                )
+                .buffer_unordered(self.io_parallelism)
+                .try_collect::<Vec<_>>()
+                .await?;
+                let list_refs = posting_lists.iter().map(|list| list.as_ref());
+                let row_ids = NGramPostingList::intersect(list_refs);
+                Ok(SearchResult::AtMost(RowIdTreeMap::from(row_ids)))
+            }
+        }
+    }
+
+    fn can_answer_exact(&self, _: &dyn AnyQuery) -> bool {
+        false
+    }
+
+    /// Load the scalar index from storage
+    async fn load(store: Arc<dyn IndexStore>) -> Result<Arc<Self>>
+    where
+        Self: Sized,
+    {
+        Ok(Arc::new(Self::from_store(store.as_ref()).await?))
+    }
+
+    /// Remap the row ids, creating a new remapped version of this index in `dest_store`
+    async fn remap(
+        &self,
+        _mapping: &HashMap<u64, Option<u64>>,
+        _dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        todo!()
+    }
+
+    /// Add the new data into the index, creating an updated version of the index in `dest_store`
+    async fn update(
+        &self,
+        _new_data: SendableRecordBatchStream,
+        _dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        todo!()
+    }
+}
+
+pub struct NGramIndexBuilder {
+    tokenizer: TextAnalyzer,
+    tokens_map: HashMap<String, u32>,
+    tokens_counter: u32,
+    bitmaps: Vec<RoaringTreemap>,
+}
+
+impl Default for NGramIndexBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl NGramIndexBuilder {
+    pub fn new() -> Self {
+        let tokenizer = NGRAM_TOKENIZER.clone();
+        Self {
+            tokenizer,
+            // Default capacity loosely based on case insensitive ascii ngrams with punctuation stripped
+            tokens_map: HashMap::with_capacity(4 * 26 * 26 * 26),
```

Comment:
where does the 4 come from? Also this assumes trigrams not ngrams I think

Reply:
I noticed some indices with more than 26 * 26 * 26 items but didn't chase it down.  Now I've realized the problem was that I wasn't considering the digits (0-9).  I've updated to 36*36*36.